File: js/clientbot/clientbot.js
/**
 */
'use strict'

// to avoid issue with the self signed certificate
// process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

const https = require('https');
const base64 = require('base-64');
const crypto = require("crypto");
const util = require('util')
var config = require('./conf_client.json');
		 
// Options to be used by request 
var options = {
		hostname: config.baseUrl,
		port: config.port,
		path: config.authUri,
		method: 'POST',
		headers: {
		    'Content-Type': 'application/x-www-form-urlencoded',
			'Authorization': 'Anonymous',
			'Origin': config.origin
		  }
	   
};

var userGUID, sessionCookie;
var pattern = /JSESSIONID=[0-9A-Fa-f]*/;

var subscribeReq = {
			"client_req_id": crypto.randomBytes(16).toString("hex"),
			"method": "SUBSCRIBE",
			"uri": "/users/me/properties"
		}

var subscribeInteractionsReq = {
		"client_req_id": crypto.randomBytes(16).toString("hex"),
		"method": "SUBSCRIBE",
		"uri": "/users/me/interactions"
	}

function postInteraction(clientReqId) {
	return {
		"client_req_id": clientReqId,
	  	"method": "POST",
	  	"uri": "/users/me/interactions",
	  	"body": {
	  			channel_type: "chat", 
	  			channel_sub_type: "text", 
				destination: config.destination,		
				attached_data: {
					"test1": 1, 
					"test2": "testing2"
			}
		}
	}
}

function putProperty(clientReqId) {
	return {
		"client_req_id": clientReqId,
	  	"method": "PUT",
	  	"uri": "/users/me/properties",
	  	"body": {	
	  		chat_address: config.originator, 
	  		alias: config.alias
	  	}		
	}
}

function messageInteraction(clientReqId, interactionGUID, message, originator) {
	return {
		"client_req_id": clientReqId,
	  	"method": "POST",
	  	"uri": "/users/me/interactions/"+interactionGUID+"/transcript/messages",
	  	"body": {	
	  		"message": message,
	  		"originator": config.originator
	  	}		
	}
}

//Authenticate user
// Callback function is used to deal with response
var callback = function(response){	
	// Continuously update stream with data
	var body = '';
	console.log('STATUS: ' + response.statusCode);
	console.log('HEADERS: ' + JSON.stringify(response.headers));
	response.setEncoding('utf8');
	response.on('data', function(data) {			
		
		body += data;
            
		// user GUID
		userGUID = JSON.parse(data).id;
		console.log('AUTHENTICATED USER: ' + userGUID);		
		// get the JSESSIONID cookie (used to idetify the session on the server side)
		sessionCookie = pattern.exec(response.headers['set-cookie'])[0]
		console.log('HTTP SESSION: ' + sessionCookie);
      
		if (sessionCookie) {
			const WebSocket = require('ws');
			// Establish websocket connection			
			const ws = new WebSocket(config.webSocketUrl, {
				origin: config.origin,
				protocolVersion: config.protocolVersion,				
				headers: {
					'Content-Type': 'application/json',
					'Cookie': sessionCookie					
				}
			});
		    
    	  ws.on('unexpected-response', function unexpResponse(request, response) {
    		  console.log(request);
    		  console.log(response);
    	  });
		
		
    	  ws.on('open', function open() {    		  
    		  console.log('WS CONNECTION ESTABLISHED: ' + config.baseUrl);
    		  console.log('');
			 
			  ws.send(JSON.stringify(putProperty(crypto.randomBytes(16).toString("hex"))));
    		  // Subscribe to resources
			  
			  ws.send(JSON.stringify(subscribeReq));
			  
    		  // Various request examples
    		  ws.send(JSON.stringify(subscribeInteractionsReq));    		  
			  ws.send(JSON.stringify(postInteraction(crypto.randomBytes(16).toString("hex"))));
			  
			  let pingTimer = setInterval(() => {
				// Send ping every 50 seconds
				ws.ping();
			 }, 5000)
    	  });
		
		  ws.on('close', function close(e) {
			  // Unsubscribe from all the resources
			  //console.log(e);
			  console.log('UNSUBSCRIBE ');
			  console.log('disconnected');
		  });

		 ws.on('ping', () => {
			console.log('got ping answering pong');
			ws.pong();
		 });
		
		 ws.on('pong', () => {
			console.log('got pong');
		 });

		  ws.on('message', function incoming(data, flags) {
			  
			  console.dir(JSON.parse(data), {depth: null, colors: true})
			  //try{
				  if (JSON.parse(data).body && JSON.parse(data).body.hasOwnProperty('transactions')) {
					  var interBody = JSON.parse(data).body;
					  for (var key in interBody) {
						  //console.log('1: '+ interBody[key]);
						  for (var trans in interBody[key]) {
							  //console.log(interBody[key][trans]);
							  for (var res in interBody[key][trans]) {
								  //console.log(interBody[key][trans].subscribed_resource);
								  if (interBody[key][trans].subscribed_resource == 'accepted') {
									  var uri = interBody[key][trans].uri;
									  var intId = interBody[key][trans].uri.split("/")[4]//interBody[key][trans][res].id;
									  if (intId && intId != null) {
										  ws.send(JSON.stringify(messageInteraction(crypto.randomBytes(16).toString("hex"), intId, config.message)));										  
									  }
								  }	  								  								  
							  }
						  }
					  }
				  }  
			  /*} catch (e) {
				  console.log(e);
			  }	*/	  		 							  		  
		  });
		
		  ws.on('error', function error(e) {
			  //console.log('error: ' + e);  	  
		  });
      }
		
   });    
   
   response.on('end', function() {
	   // Data received completely.
	   console.log(body);
   });
   
   response.on('error', function(e){
	   console.log("Request error: " + e); 
   });
}
// Make a request to the server
console.log(options)
var res = https.request(options, callback);
res.end();

File: py/clientbot/clientbot.py
import asyncio
import pathlib
import ssl
import websockets
import requests
import configparser
import base64
import uuid
import json
# %%
ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
ssl_context.load_default_certs()
from urllib import request, parse
#basic = uid + ":" + pwd
basic = "Anonymous"
basic_bytes = basic.encode("ascii") 
b64_bytes = base64.b64encode(basic_bytes) 
b64_str = "Basic " + b64_bytes.decode("ascii")
print(b64_str)
b = b64_str
data = parse.urlencode({"Authorization": b}).encode()
heads = {
"Content-Type": "application/x-www-form-urlencoded",
"Origin": "prod-eu.sapcctr.com",
"Authorization": "Anonymous"
}

#req =  request.Request("https://prod-eu.sapcctr.com/standarddemo/ecfs/authentication", headers=heads,  data=data) # this will make the method "POST"
#request.urlopen(req)
resp = requests.post('https://prod-eu.sapcctr.com/standarddemo/visitor/ecfs/authentication', headers=heads, data=data) 
#print(resp.cookies)
print (requests.utils.dict_from_cookiejar(resp.cookies))
dictestives = requests.utils.dict_from_cookiejar(resp.cookies)
jsessionid = dictestives["JSESSIONID"]
jsessionstr = "JSESSIONID=" + jsessionid
cookiehead  = [("Cookie", jsessionstr)]

alias = "Visitor"
visitor = input("Give your visitor email address: ")
destination = input("Give queue address: ")

# Put user properties. You should set users email address with this message
put_user_properties = {
    "client_req_id": uuid.uuid4().hex,
	"method": "PUT",
	"uri": "/users/me/properties",
	"body": {	
	  		"chat_address": visitor, 
	  		"alias": alias
    }
}

# Subscribe to interactions
subscribe_me = {
	"client_req_id": uuid.uuid4().hex,
	"method": "SUBSCRIBE",
	"uri": "/users/me/properties"
}

# Subscribe to my interactions
subscribe_my_interactions = {
	"client_req_id": uuid.uuid4().hex,
	"method": "SUBSCRIBE",
	"uri": "/users/me/interactions"
}

# Post interaction
post_interaction = {
    "client_req_id": uuid.uuid4().hex,
    "method": "POST",
    "uri": "/users/me/interactions",
    "body": {
	    "channel_type": "chat", 
	    "channel_sub_type": "text", 
	    "destination": destination,		
	    "attached_data": {
    	    "foo": "bar", 
		    "bin": "baz",
		    "mycreature": {
                "head": "0", "body": "X", "legs":"||"
            }
	    }
    }
}

# Post messages to chat

is_running = True

async def openseasam():
    uri = "wss://prod-eu.sapcctr.com/standarddemo/visitor/ecfs/ws_endpoint/"
    is_connected = False
    async with websockets.connect(
        uri, 
        ssl=ssl_context,
        extra_headers=cookiehead, 
        subprotocols="chat"
    ) as websocket:
        while True:
            try:              
                message = await websocket.recv()
                parsed = json.loads(message)
                print(json.dumps(parsed, indent=4, sort_keys=True))

                if (is_connected == False):
                    print("putting properties...")
                    await websocket.send(json.dumps(put_user_properties))
                    print("subscribe me")
                    await websocket.send(json.dumps(subscribe_me))
                    print("subscribing to my interactions...")
                    await websocket.send(json.dumps(subscribe_my_interactions))
                    print("sending interaction")
                    await websocket.send(json.dumps(post_interaction))
                    is_connected=True

                
            except websockets.exceptions.ConnectionClosed:
                print('ConnectionClosed')
                is_connected = False
                is_running = False
                break

asyncio.get_event_loop().run_until_complete(asyncio.wait([   
   openseasam()
]))



# %%
